tokio/src/runtime/io/registration_set.rs

```diff
@@ -0,0 +1,134 @@
+use crate::loom::sync::atomic::AtomicUsize;
+use crate::runtime::io::ScheduledIo;
+use crate::util::linked_list::{self, LinkedList};
+
+use std::io;
+use std::ptr::NonNull;
+use std::sync::atomic::Ordering::{Acquire, Release};
+use std::sync::Arc;
+
+pub(super) struct RegistrationSet {
+    num_pending_release: AtomicUsize,
+}
+
+pub(super) struct Synced {
+    // True when the I/O driver shutdown. At this point, no more registrations
+    // should be added to the set.
+    is_shutdown: bool,
+
+    // List of all registrations tracked by the set
+    registrations: LinkedList<Arc<ScheduledIo>, ScheduledIo>,
+
+    // Registrations that are pending drop. When a `Registration` is dropped, it
+    // stores its `ScheduledIo` in this list. The I/O driver is responsible for
+    // dropping it. This ensures the `ScheduledIo` is not freed while it can
+    // still be included in an I/O event.
+    pending_release: Vec<Arc<ScheduledIo>>,
+}
```

Comment:
what's the motivation for not just having this be 
```suggestion
pub(super) struct RegistrationSet {
    num_pending_release: AtomicUsize,
    synced: Mutex<Synced>,
}

struct Synced {
    // True when the I/O driver shutdown. At this point, no more registrations
    // should be added to the set.
    is_shutdown: bool,

    // List of all registrations tracked by the set
    registrations: LinkedList<Arc<ScheduledIo>, ScheduledIo>,

    // Registrations that are pending drop. When a `Registration` is dropped, it
    // stores its `ScheduledIo` in this list. The I/O driver is responsible for
    // dropping it. This ensures the `ScheduledIo` is not freed while it can
    // still be included in an I/O event.
    pending_release: Vec<Arc<ScheduledIo>>,
}
``` 
and changing all the methods that take a `&self, &mut Synced` to just lock `self.synced` internally? as far as I could tell, there's nowhere where we lock `synced` for multiple operations, unless i can't read and missed something?

Reply:
I plan on unifying mutexes in a later PR.